File: behavioral/iterator/ruby/iterator.rb
class Lion
  attr_reader :weight, :age

  def initialize(weight, age)
    @weight = weight
    @age = age
  end
end

class Tiger
  attr_reader :weight, :age

  def initialize(weight, age)
    @weight = weight
    @age = age
  end
end

class Bear
  attr_reader :weight, :age

  def initialize(weight, age)
    @weight = weight
    @age = age
  end
end

class OhMy
  attr_reader :weight, :age

  def initialize(weight, age)
    @weight = weight
    @age = age
  end
end

class Zookeeper
  # include Enumerable
  def initialize(animals = [])
    @animals = animals
    @index = 0
  end

  # def each(&block)
  #   @animals.each(block)
  # end

  def has_next?
    @index < @animals.length
  end

  def animal
    @animals[index]
  end

  def next_animal
    value = @animals[@index]
    @index += 1
    value
  end

  def map
    mapped = []
    while has_next? do
      mapped << yield(next_animal)
    end
    mapped
  end
end

animals = [
  Lion.new(100, 3),
  Tiger.new(200, 9),
  Bear.new(700, 5),
  OhMy.new(300, 2),
]

puts Zookeeper.new(animals).map{ |animal| "weight: #{animal.weight}"}
puts Zookeeper.new(animals).map{ |animal| "age: #{animal.age}"}

File: behavioral/iterator/iterator.js
// http://eloquentjavascript.net/06_object.html

class Lion {
  constructor(weight, age) {
    this.weight = weight
    this.age = age
  }
}

class Tiger {
  constructor(weight, age) {
    this.weight = weight
    this.age = age
  }
}

class Bear {
  constructor(weight, age) {
    this.weight = weight
    this.age = age
  }
}

class OhMy {
  constructor(weight, age) {
    this.weight = weight
    this.age = age
  }
}

class Zoo {
  constructor(animals) {
    this.animals = animals
  }
}

class Zookeeper {
  constructor(zoo) {
    this.i = 0;
    this.zoo = zoo;
    console.log("")
  }

  next() {
    if (this.i >= this.zoo.animals.length) return {done: true};

    let value = {
      i: this.i,
      value: this.zoo.animals[this.i]
    };

    this.i++;
    return {value, done: false};
  }
}

Zoo.prototype[Symbol.iterator] = function() {
  return new Zookeeper(this);
};

let animals = [
  new Lion(100, 3),
  new Tiger(200, 9),
  new Bear(700, 5),
  new OhMy(300, 2),
]

let zoo = new Zoo(animals);
for (let {i, value} of zoo) {
  console.log(i, value);
}
